### DISCUSSION TITLE
Performance/overhead with disabled pthread vs. pthread but single threaded

### DISCUSSION THREAD

**Discussion by svenpilz** [emscripten-core/emscripten#15914] [Q&A] [answered]
Hi fellow Emscripten users,

we are trying to make use of multi threading in an application which gets called from JS, does some longer running calculations (think of a big for loop) and returns the result back to JS. Now we observed that just enabling _pthreads_ (we also use _simd_, `-msimd128 -pthread`) without actually using multiple threads (our dispatcher falls back to single threaded execution in that case), the runtime increases by maybe 10% (about 2sec here). Also for parts of the code that don't know about threading.

Is this expected? Is there a rather big overhead for just having _pthread_ support activated? We don't have exception support enabled and it seems there is no difference between static/dynamic memory size.

Thanks
Sven

**Comment by kripken** [comment, marked as answer]
Just enabling pthreads has the effect of making various things use actual locks (like malloc, file operations, etc.). Also atomic operations are actually atomic. Both of those slowdowns happen even if you never create another thread.

A 10% slowdown is more than I'd expect, though, so it might be interesting to profile in the browser devtools and see how the profiles differ between the two builds - that diff could show interesting things.

**Comment by svenpilz** [comment]
Thanks for the tip and your time. As you thought, seems it is `malloc`/`free` and their _atomics_. They differ by about 2sec total time. Guess we should reduce heap allocations a bit :). The code uses a lot of dynamic _Eigen_ matrices.

## With `pthread` and run with a single thread in a web worker.

| Self Time            |  Total Time           | Activity     |
| ---------------------: | ----------------------: | :------------ |
| 850.0 ms, 4.8 % | 850.0 ms, 4.8 %   | a_swap.1 |
| 687.1 ms, 3.9 % | 1684.8 ms, 9.5 % | dlfree        |
| 684.0 ms. 3.9 % | 684.0 ms, 3.9 %   | a_cas.3     |
| 599.3 ms. 3.4 % | 1622.1 ms, 9.1 % | dlmalloc    |

## Without `pthread` in web worker.
| Self Time            |  Total Time         | Activity |
| ---------------------: | --------------------: | :---------- |
| 591.9 ms, 4.0 % | 591.9 ms, 4.0 % | dlfree     |
| 490.4 ms, 3.3 % | 491.0 ms, 3.3 % | dlmalloc |

**Reply by sbc100** [reply]
Interesting.  I wonder if using `tcmalloc` or something like that might make sense?

**Comment by Mintyboi** [comment]
@kripken I understand this discussion has been resolved. Please let me know if I should start another topic.

I just recently enabled PThreads in my application and am currently hitting into the above performance degradation. Just wanted to check if there's anything that I can do or try out.
Is the tcmalloc approach mentioned by @sbc100 something viable?

**Reply by sbc100** [reply]
@Mintyboi, are you able to profile to determine if the allocator is the reason for your slowdown, or something else?   I've not tried tcmalloc myself but it should be easy to compile it into your application (you can build with `-sMALLOC=none` to be sure the default one is not linked in).

**Reply by Mintyboi** [reply]
Yes, I have snapshot of the chrome profiler that I can attach later.
Sure, I'll give that a try to see if I can link in tcmalloc.
Thank you

**Reply by svenpilz** [reply]
I would also be interested in your findings.

We could remove most bottlenecks by
* Limiting us to matrices with static number of columns (either because we only needed them or by using templates).
* Avoiding allocations in loops by having one big chunk of memory outside.
* For functions like `std::vector<T> findThings()` have also one `T findFirstThing()` that avoids allocation.

It made our code more explicit, but maybe that is actually a good thing. But we also found usage of `std::vector` with static sizes that were just created to call functions `foo(const std::vector<>&)`, which we replaced by `std::array` and made the argument a `span`. Which I would consider as a real improvement.

**Reply by svenpilz** [reply]
Regarding the profiling. I recently noted that Chrome somehow had issues profiling all spawned workers. Firefox was fine, once I enabled the detection of new threads. Have you noted something similar? While you are profiling, is your code also faster on Firefox compared to Chrome? We see significant performance differences there.

**Reply by svenpilz** [reply]
@Mintyboi, did you had any luck getting `tcmalloc` up and running? I tried it today, but I'm still struggling as a `bazel` newbie and the timebox I have for playing with such things :).

**Reply by svenpilz** [reply]
@sbc100, do you know if `mimalloc` should in theory work under `Emscripten`?

**Reply by svenpilz** [reply]
As the background, we had some performance issues with some multi-threaded code that needs a priority queue and it seems to get stalled by waiting.

**Reply by sbc100** [reply]
I don't see any reason why tcmalloc wouldn't work with emscripten.

**Reply by svenpilz** [reply]
So far I was not successful. `tcmalloc` would not compile (https://github.com/google/tcmalloc/blob/d679038aee140b22dc582b179b9774e2695e1c87/tcmalloc/internal/config.h#L93 as well as https://github.com/google/tcmalloc/blob/d679038aee140b22dc582b179b9774e2695e1c87/tcmalloc/internal/config.h#L101) and `gperftools` fails with "mismatch between intptr_t and the Atomic32 or Atomic64" (tried to patch a little).

But even if I get `libtcmalloc_minimal` somehow to compile, linking the application fails with 

```
wasm-ld: error: /*/wasm32-emscripten/crtbegin.o: undefined symbol: emscripten_builtin_memalign
wasm-ld: error: /*/wasm32-emscripten/libc-mt.a(__environ.o): undefined symbol: emscripten_builtin_malloc
wasm-ld: error: /*/wasm32-emscripten/libc-mt.a(__environ.o): undefined symbol: emscripten_builtin_malloc
wasm-ld: error: /*/wasm32-emscripten/crtbegin.o: undefined symbol: emscripten_builtin_free
wasm-ld: error: /*/wasm32-emscripten/crtbegin.o: undefined symbol: emscripten_builtin_free
wasm-ld: error: /*/wasm32-emscripten/crtbegin.o: undefined symbol: emscripten_builtin_memalign
wasm-ld: error: /*/wasm32-emscripten/libc-mt.a(__environ.o): undefined symbol: emscripten_builtin_malloc
wasm-ld: error: /*/wasm32-emscripten/crtbegin.o: undefined symbol: emscripten_builtin_free
wasm-ld: error: /*/wasm32-emscripten/libc-mt.a(__environ.o): undefined symbol: emscripten_builtin_malloc
wasm-ld: error: /*/wasm32-emscripten/crtbegin.o: undefined symbol: emscripten_builtin_free
wasm-ld: error: /*/wasm32-emscripten/libc-mt.a(locale_map.o): undefined symbol: __libc_malloc
wasm-ld: error: /*/wasm32-emscripten/libc-mt.a(locale_map.o): undefined symbol: __libc_malloc
```

Any ideas how to go from here?

**Reply by sbc100** [reply]
These are internal aliases for the malloc functions, you ned to add these aliases to tcmalloc, like we do for emmalloc and dlmalloc.  e.g. https://github.com/emscripten-core/emscripten/blob/a0d43fc3a07fa9161b41d030036695c24d25cb30/system/lib/emmalloc.c#L829-L830

**Reply by svenpilz** [reply]
> @sbc100, do you know if `mimalloc` should in theory work under `Emscripten`?

I got `mimalloc` to compile with 

```cpp
  void* __libc_malloc(size_t size)                      MI_FORWARD1(mi_malloc, size)
  void* __libc_calloc(size_t count, size_t size)        MI_FORWARD2(mi_calloc, count, size)
  void* __libc_realloc(void* p, size_t size)            MI_FORWARD2(mi_realloc, p, size)
  void  __libc_free(void* p)                            MI_FORWARD0(mi_free, p)
  void* __libc_memalign(size_t alignment, size_t size)  { return mi_memalign(alignment, size); }

// Added 

  void* emscripten_builtin_malloc(size_t size)                      MI_FORWARD1(mi_malloc, size)
  void  emscripten_builtin_free(void* p)                            MI_FORWARD0(mi_free, p)
  void* emscripten_builtin_memalign(size_t alignment, size_t size)  { return mi_memalign(alignment, size); }     
```

but it crashes on runtime with multi-threading.

**Reply by svenpilz** [reply]
Ah, that seems be design: https://github.com/microsoft/mimalloc/issues/720#issuecomment-1499786394.

**Reply by svenpilz** [reply]
Will try `tcmalloc` again once I get some time next in the next sprints :). I didn't get it to export `malloc`/`free`, not sure why.

**Reply by Mintyboi** [reply]
Sorry for the late reply, I did not try out tcmalloc. Found out about [rpmalloc](https://github.com/mjansson/rpmalloc) and was giving it a go just as a sort of POC to understand these issues but had some issues integrating into my application and did not had the time to continue in a small sample app.

@mannk123 did prototyped using mspace_malloc successfully https://github.com/emscripten-core/emscripten/blob/90d76db21200d42f5536bc3a01a22303e398d8b4/system/lib/dlmalloc.c#L5604 and found that it improves performance by 5x, from ~500ms to ~100ms for millions of malloc + free operations
I'll let him share his impl details and sample app.
